viewing_party/party.py

```diff
@@ -0,0 +1,254 @@
+                return genre
+
+def get_users_watched_movies(user_data):
+
+    users_watched_movies = set()
+    watched_movie_details = user_data["watched"]
+    
+    # loop over the user's list of watched movies
+    for movie in watched_movie_details:
+        movie_title = movie["title"]
+        # add the movie title to a set of watched movies
+        users_watched_movies.add(movie_title)
+
+    # return set of user's watched movies 
+    return users_watched_movies
+
+def get_friends_watched_movies(user_data):
+
+    movies_watched_all_friends = set()
+    friends_list = user_data["friends"]
+
+    # loop over a list of the user's friends
+    for friend in friends_list:
+        friends_movies = friend["watched"]
+        # loop over each movie in the friend's watched movie list
+        for movie in friends_movies:
+            movie_title = movie["title"]
+            # add the movie title to a set of all the movies watched by the user's friends
+            movies_watched_all_friends.add(movie_title)
+
+    # return set of the movies watched by user's friends
+    return movies_watched_all_friends
+
+def get_unique_watched(user_data):
+
+    # get a list of the user's watched movies
+    users_movies = get_users_watched_movies(user_data)
+    # get a list of the user's friend's watched movies
+    friends_movies = get_friends_watched_movies(user_data)
+
+    # get the difference between the user's movies and the user's friend's movies
+    unique_movies = users_movies.difference(friends_movies)
+    
+    users_unique_movies = []
+    
+    # loop over the list of movies only the user has watched
+    for movie in unique_movies:
+        movie_title = {"title":movie}
+        # add the movie title to the list of unique movies
+        users_unique_movies.append(movie_title)
+
+    # return the list of movies only the user has watched
+    return users_unique_movies
+
+def get_friends_unique_watched(user_data):
+
+    # get a list of the user's watched movies
+    users_movies = get_users_watched_movies(user_data)
+    # get a list of the user's friend's watched movies
+    friends_movies = get_friends_watched_movies(user_data)
+
+    # get the difference between the user's friend's movies and the user's movies
+    unique_movies = friends_movies.difference(users_movies)
+    
+    friends_unique_movies = []
+
+    # loop over the list of unique movies the user's friends have watched
+    for movie in unique_movies:
+        movie_title = {"title": movie}
+        # add the movie title to the list of unique movies
+        friends_unique_movies.append(movie_title)
+
+    # return the list of movies the user's friends have watched but the user has not
+    return friends_unique_movies
+
+def get_friends_unique_watched_with_details(user_data):
+
+    # get a list of movie only theuser's friends have watched
+    users_unwatched_movies = get_friends_unique_watched(user_data)
```

Comment:
If you flattened this down to a set of titles, you could have O(1) lookups of whether a movie's title is one of the unique  titles the user hasn't seen. Rather than starting the subsequent iteration from the unwatched list, we could start it on the friends, and check whether the title of each film is in the unique set. This could let you return a list of full movie dicts later on.